toolchain/check/function.cpp

```diff
@@ -0,0 +1,216 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/check/function.h"
+
+namespace Carbon::Check {
+
+// Returns true if the function signature has an error, which will have
+// previously been diagnosed.
+static auto FunctionSignatureHasError(Context& context,
+                                      const SemIR::Function& fn) -> bool {
+  if (fn.return_type_id == SemIR::TypeId::Error) {
+    return true;
+  }
+  for (auto param_refs_id : {fn.implicit_param_refs_id, fn.param_refs_id}) {
+    if (param_refs_id != SemIR::InstBlockId::Empty) {
+      for (auto param_id : context.inst_blocks().Get(param_refs_id)) {
+        if (context.insts().Get(param_id).type_id() == SemIR::TypeId::Error) {
+          return true;
+        }
+      }
+    }
+  }
+  return false;
+}
+
+// Returns true if a param agrees. The caller is expected to provide a
+// diagnostic.
+// TODO: Consider moving diagnostics here, particularly to differentiate
+// between type and name mistakes. For now, taking the simpler approach because
+// I also think we may want to refactor params.
+static auto DoesParamAgree(Context& context, SemIR::InstId new_param_ref_id,
+                           SemIR::InstId prev_param_ref_id) -> bool {
+  auto new_param_ref = context.insts().Get(new_param_ref_id);
+  auto prev_param_ref = context.insts().Get(prev_param_ref_id);
+  if (new_param_ref.kind() != prev_param_ref.kind() ||
+      new_param_ref.type_id() != prev_param_ref.type_id()) {
+    return false;
+  }
+
+  if (new_param_ref.Is<SemIR::AddrPattern>()) {
+    new_param_ref =
+        context.insts().Get(new_param_ref.As<SemIR::AddrPattern>().inner_id);
+    prev_param_ref =
+        context.insts().Get(prev_param_ref.As<SemIR::AddrPattern>().inner_id);
+  }
+
+  if (new_param_ref.kind() != prev_param_ref.kind()) {
+    return false;
+  }
+
+  if (new_param_ref.Is<SemIR::AnyBindName>()) {
+    new_param_ref =
+        context.insts().Get(new_param_ref.As<SemIR::AnyBindName>().value_id);
+    prev_param_ref =
+        context.insts().Get(prev_param_ref.As<SemIR::AnyBindName>().value_id);
+  }
+
+  CARBON_CHECK(new_param_ref.Is<SemIR::Param>()) << new_param_ref;
+  CARBON_CHECK(prev_param_ref.Is<SemIR::Param>()) << prev_param_ref;
+
```

Comment:
Do these produce better diagnostics than the ones we get from `As`? If so, would it make sense to improve the CHECK there instead of adding one here?

Reply:
Removed as redundant. Not investigating diagnostics.

Reply:
Removed as redundant, I don't think I looked at diagnostics, this was just not cleaned up.